### PR TITLE
Test perf of Request.Body.CopyToAsync()

### PR DESCRIPTION
**Pull request by halter73**
@mikeharder @davidfowl @cesarbs

### CODE REVIEW

**Comment by halter73** [comment]
This was originally made to test the perf of Kestrel's CopyToAsync changes. I used the default post.lua script and the /copy endpoint to test it. The RPS numbers are well withing a stddev on my laptop, but the number of gen 0 collections is far less (not that gen 0 collections are very expensive obvs).

## Before

    C:\Users\shalter\dev\benchmarks\src\Benchmarks [dev +2 ~6 -0 !]> #dev-to-body
    C:\Users\shalter\dev\benchmarks\src\Benchmarks [dev +2 ~6 -0 !]> dotnet run -c Release --scenarios copy --server.urls http://*:5000
    ASP.NET Core Benchmarks
    -----------------------
    Current directory: C:\Users\shalter\dev\benchmarks\src\Benchmarks
    Scenario configuration found in scenarios.json and/or command line args

    The following scenarios were enabled:
      Copy -> /copy

    Using server Kestrel
    Server GC is currently ENABLED
    Press 'C' to force GC or any other key to display GC stats

    Hosting environment: Production
    Content root path: C:\Users\shalter\dev\benchmarks\src\Benchmarks
    Now listening on: http://*:5000
    Application started. Press Ctrl+C to shut down.

    Allocated: 19.00 MB
    Gen 0: 1949, Gen 1: 44, Gen 2: 3

    Forcing GC... done!

    Allocated: 3.29 MB
    Gen 0: 1951, Gen 1: 46, Gen 2: 5
    Application is shutting down...

    [shalter@shalter-arch wrk]$ #dev-to-body
    [shalter@shalter-arch wrk]$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://win8:5000/copy
    Running 10s test @ http://win8:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    22.31ms   50.79ms 499.99ms   95.81%
        Req/Sec   626.83    112.44     0.88k    86.77%
      191754 requests in 10.04s, 28.16MB read
    Requests/sec:  19108.19
    Transfer/sec:      2.81MB
    [shalter@shalter-arch wrk]$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://win8:5000/copy
    Running 10s test @ http://win8:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    12.82ms   15.33ms 254.07ms   98.57%
        Req/Sec   695.38    174.38     1.40k    83.36%
      201116 requests in 10.03s, 29.54MB read
    Requests/sec:  20053.62
    Transfer/sec:      2.95MB
    [shalter@shalter-arch wrk]$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://win8:5000/copy
    Running 10s test @ http://win8:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    14.22ms   16.96ms 299.23ms   98.27%
        Req/Sec   638.15     86.92     0.85k    79.90%
      201922 requests in 10.03s, 29.66MB read
    Requests/sec:  20141.49
    Transfer/sec:      2.96MB

## After

    C:\Users\shalter\dev\benchmarks\src\Benchmarks [dev +2 ~6 -0 !]> #copyless-to-body
    C:\Users\shalter\dev\benchmarks\src\Benchmarks [dev +2 ~6 -0 !]> dotnet run -c Release --scenarios copy --server.urls http://*:5000
    ASP.NET Core Benchmarks
    -----------------------
    Current directory: C:\Users\shalter\dev\benchmarks\src\Benchmarks
    Scenario configuration found in scenarios.json and/or command line args

    The following scenarios were enabled:
      Copy -> /copy

    Using server Kestrel
    Server GC is currently ENABLED
    Press 'C' to force GC or any other key to display GC stats

    Hosting environment: Production
    Content root path: C:\Users\shalter\dev\benchmarks\src\Benchmarks
    Now listening on: http://*:5000
    Application started. Press Ctrl+C to shut down.

    Allocated: 19.01 MB
    Gen 0: 12, Gen 1: 3, Gen 2: 1

    Forcing GC... done!

    Allocated: 2.91 MB
    Gen 0: 14, Gen 1: 5, Gen 2: 3
    Application is shutting down...

    [shalter@shalter-arch wrk]$ #copyless-to-body
    [shalter@shalter-arch wrk]$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://win8:5000/copy
    Running 10s test @ http://win8:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    15.16ms   22.46ms 316.03ms   97.75%
        Req/Sec   644.73    126.47     1.08k    79.07%
      197206 requests in 10.08s, 28.96MB read
    Requests/sec:  19566.71
    Transfer/sec:      2.87MB
    [shalter@shalter-arch wrk]$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://win8:5000/copy
    Running 10s test @ http://win8:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    14.62ms   19.43ms 305.61ms   97.90%
        Req/Sec   643.31    101.32     0.94k    78.96%
      203276 requests in 10.03s, 29.85MB read
    Requests/sec:  20276.71
    Transfer/sec:      2.98MB
    [shalter@shalter-arch wrk]$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://win8:5000/copy
    Running 10s test @ http://win8:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    14.21ms   18.05ms 280.33ms   97.98%
        Req/Sec   652.34    120.25     1.33k    85.88%
      204888 requests in 10.03s, 30.09MB read
    Requests/sec:  20431.49
    Transfer/sec:      3.00MB

**Comment by halter73** [comment]
https://github.com/aspnet/KestrelHttpServer/pull/1105

**Comment by halter73** [comment]
The difference is 1949 vs 12 gen 0 collections, 44 vs 3  gen 1 collections and 3 vs 1 gen 2 collections before forcing any collections.

**Comment by davidfowl** [comment]
👏

**Comment by cesarbs** [comment]
:shipit:

File: src/Benchmarks/Properties/launchSettings.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
       }
     },
     "Benchmarks": {
-      "commandName": "Benchmarks",
+      "commandName": "Project",
```
Comment (mikeharder) [inline src/Benchmarks/Properties/launchSettings.json]: Is this change intended?  What is the purpose?

**Comment by halter73** [comment]
Seeing nearly a 4x improvement in RPS when we run this benchmark in our perflab with pipelining! 🎉 

## Before

    C:\code\github\aspnet\benchmarks\src\Benchmarks>dotnet run -c Release --scenarios copy --server.urls http://*:5000
    ASP.NET Core Benchmarks
    -----------------------
    Current directory: C:\code\github\aspnet\benchmarks\src\Benchmarks
    Scenario configuration found in scenarios.json and/or command line args

    The following scenarios were enabled:
      Copy -> /copy

    Using server Kestrel
    Server GC is currently ENABLED
    Press 'C' to force GC or any other key to display GC stats

    Hosting environment: Production
    Content root path: C:\code\github\aspnet\benchmarks\src\Benchmarks
    Now listening on: http://*:5000
    Application started. Press Ctrl+C to shut down.

    Allocated: 106.49 MB
    Gen 0: 2313, Gen 1: 137, Gen 2: 7

    Forcing GC... done!

    Allocated: 5.63 MB
    Gen 0: 2315, Gen 1: 139, Gen 2: 9
    Application is shutting down...

    adx@adx-3u-svr02:~/code/github/wg/wrk$ #dev
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 16
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    16.91ms   17.48ms 185.50ms   89.28%
        Req/Sec     6.49k     1.16k   11.37k    86.32%
      2003141 requests in 10.10s, 294.19MB read
      Socket errors: connect 0, read 0, write 300, timeout 0
    Requests/sec: 198313.55
    Transfer/sec:     29.13MB
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 16
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    16.09ms   15.91ms 249.78ms   88.76%
        Req/Sec     6.52k     1.20k   19.74k    86.37%
      2046536 requests in 10.10s, 300.57MB read
      Socket errors: connect 0, read 0, write 300, timeout 0
    Requests/sec: 202629.87
    Transfer/sec:     29.76MB
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 16
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    16.10ms   16.50ms 224.28ms   89.41%
        Req/Sec     6.55k     1.28k   22.52k    85.54%
      2040074 requests in 10.10s, 299.62MB read
      Socket errors: connect 0, read 0, write 300, timeout 0
    Requests/sec: 201982.69
    Transfer/sec:     29.66MB


## After

    C:\code\github\aspnet\benchmarks\src\Benchmarks>dotnet run -c Release --scenarios copy --server.urls http://*:5000
    ASP.NET Core Benchmarks
    -----------------------
    Current directory: C:\code\github\aspnet\benchmarks\src\Benchmarks
    Scenario configuration found in scenarios.json and/or command line args

    The following scenarios were enabled:
      Copy -> /copy

    Using server Kestrel
    Server GC is currently ENABLED
    Press 'C' to force GC or any other key to display GC stats

    Hosting environment: Production
    Content root path: C:\code\github\aspnet\benchmarks\src\Benchmarks
    Now listening on: http://*:5000
    Application started. Press Ctrl+C to shut down.

    Allocated: 79.96 MB
    Gen 0: 90, Gen 1: 19, Gen 2: 2

    Forcing GC... done!

    Allocated: 5.59 MB
    Gen 0: 92, Gen 1: 21, Gen 2: 4
    Application is shutting down...

    adx@adx-3u-svr02:~/code/github/wg/wrk$ #copyless
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 16
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    14.69ms   32.19ms 293.26ms   89.46%
        Req/Sec    25.18k     4.13k   44.36k    79.88%
      7780714 requests in 10.10s, 1.12GB read
      Socket errors: connect 0, read 0, write 300, timeout 0
    Requests/sec: 770481.24
    Transfer/sec:    113.16MB
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 16
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    13.84ms   32.50ms 350.60ms   89.70%
        Req/Sec    25.27k     3.69k   47.68k    75.79%
      7999287 requests in 10.10s, 1.15GB read
      Socket errors: connect 0, read 0, write 300, timeout 0
    Requests/sec: 791996.46
    Transfer/sec:    116.32MB
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 16
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    12.35ms   30.56ms 388.73ms   91.04%
        Req/Sec    24.98k     3.47k   74.09k    81.38%
      7916825 requests in 10.10s, 1.14GB read
      Socket errors: connect 0, read 0, write 600, timeout 0
    Requests/sec: 783893.03
    Transfer/sec:    115.13MB

I'm going to add the pipelined post.lua I used to the PR.

File: scripts/post.lua [hunk 2]
```diff
@@ -0,0 +1,22 @@
+local pipelineDepth = 15
```
Comment (mikeharder) [inline scripts/post.lua]: Typical default for pipeline depth is 16

File: scripts/post.lua [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+function init(args)
+   wrk.method = "POST"
+   wrk.body   = "foo=bar&baz=quux"
+   wrk.headers["Content-Type"] = "application/x-www-form-urlencoded"
+
+   if args[1] ~= nil then
+      pipelineDepth = args[1]
+   end
+
```
Comment (mikeharder) [inline scripts/post.lua]: The script should print the pipelineDepth like `pipeline.lua`, so its included in the output.

**Comment by halter73** [comment]
About a 2x improvement without pipelining:


## Before
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 1
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     2.52ms    7.27ms 141.25ms   97.65%
        Req/Sec     5.01k     0.85k    8.83k    90.98%
      1550953 requests in 10.10s, 227.78MB read
      Socket errors: connect 0, read 0, write 489, timeout 0
    Requests/sec: 153564.10
    Transfer/sec:     22.55MB
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 1
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     1.90ms    1.98ms  70.69ms   91.72%
        Req/Sec     5.02k   437.31     8.73k    83.21%
      1592348 requests in 10.10s, 233.86MB read
      Socket errors: connect 0, read 0, write 300, timeout 0
    Requests/sec: 157673.33
    Transfer/sec:     23.16MB


## After

    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 1
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     1.85ms    8.30ms 142.05ms   97.64%
        Req/Sec    10.38k     1.51k   17.12k    91.58%
      3219496 requests in 10.10s, 472.83MB read
      Socket errors: connect 0, read 0, write 536, timeout 0
    Requests/sec: 318763.24
    Transfer/sec:     46.82MB
    adx@adx-3u-svr02:~/code/github/wg/wrk$ ./wrk -c 256 -t 32 -d 10 -s post.lua http://10.0.0.100:5000/copy -- 1
    Running 10s test @ http://10.0.0.100:5000/copy
      32 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     2.60ms    9.18ms 160.04ms   95.40%
        Req/Sec    10.22k   839.38    13.82k    65.95%
      3285931 requests in 10.10s, 482.59MB read
    Requests/sec: 325363.86
    Transfer/sec:     47.78MB

File: src/Benchmarks/Properties/launchSettings.json (hunk 1 above)
Comment (halter73) [inline src/Benchmarks/Properties/launchSettings.json]: Yes. F5 doesn't work without this change.

File: scripts/post.lua (hunk 2 above)
Comment (halter73) [inline scripts/post.lua]: It's 15 in the sample pipeline.lua from wrk. 1 in the pipeline.lua in the same dir as this in the repo.

File: scripts/post.lua (hunk 3 above)
Comment (halter73) [inline scripts/post.lua]: It's soo noisy though since it's printed once per thread. Annoying when you want to copy results into github comments.

**Comment by benaadams** [comment]
Boom! 🚀

File: scripts/post.lua (hunk 2 above)
Comment (mikeharder) [inline scripts/post.lua]: I would default to 1 then.

File: scripts/post.lua (hunk 3 above)
Comment (mikeharder) [inline scripts/post.lua]: I still think we need this, otherwise you have no idea if the correct pipeline level was used.

**Comment by mikeharder** [comment]
I'm seeing similar improvements in Azure (with higher pipeline levels).  With pipeline=64, RPS improves from 147k to 372k.

File: src/Benchmarks/Middleware/CopyToAsyncMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public Task Invoke(HttpContext httpContext)
+        {
+            if (httpContext.Request.Path.StartsWithSegments(_path, StringComparison.Ordinal))
+            {
+                httpContext.Response.StatusCode = 200;
+                httpContext.Response.ContentType = "text/plain";
+
+                return httpContext.Request.Body.CopyToAsync(httpContext.Response.Body);
```
Comment (mikeharder) [inline src/Benchmarks/Middleware/CopyToAsyncMiddleware.cs]: If I do a GET request to `/copy`, it works if the headers do not contain "Connection: Close".  However, if the headers include "Connection: Close", this method never returns.  Is this a bug or by design?  For other handlers like `/plaintext`, GET requests work fine with or without "Connection: Close".

Comment (mikeharder) [inline src/Benchmarks/Middleware/CopyToAsyncMiddleware.cs]: POST with a body does work with `Connection: Close`.

Comment (halter73) [inline src/Benchmarks/Middleware/CopyToAsyncMiddleware.cs]: POST works with a `Connection: Close` because there is also a content-length. For GETs, we should assume the content-length is 0 unless something else is explicitly specified like we do for normal keep-alive requests.

This is a bug: https://github.com/aspnet/KestrelHttpServer/issues/1104
